improver/utilities/cube_extraction.py

```diff
@@ -352,16 +353,126 @@ def pressure_grid(variable_on_pressure: Cube) -> np.ndarray:
+        )
+        cube.data = data
+        if not (
+            np.logical_or(np.isnan(cube.data), np.isinf(cube.data)).any()
+            or np.ma.is_masked(cube.data)
+        ):
+            # We've filled everything in. No need to try again.
+            return
+        self._one_way_fill(
+            data, pressure_axis, pressure_points, v_increment, reverse=False
+        )
+        cube.data = data
+
+    def _one_way_fill(
+        self,
+        data: np.ma.MaskedArray,
+        pressure_axis: int,
+        pressure_points: np.ndarray,
+        v_increment: np.ndarray,
+        reverse: bool = False,
+    ):
+        """
+        Scans through the pressure axis forwards or backwards, filling any missing data with the
+        value plus (or minus in reverse) the specified increment.
+        """
+        last_p_slice = [slice(None) for _ in range(data.ndim)]
+        if reverse:
+            local_increment = -v_increment
+            first = len(pressure_points) - 2
+            last = -1
+            step = -1
+            last_p_slice[pressure_axis] = slice(
+                len(pressure_points) - 1, len(pressure_points)
+            )
+        else:
+            local_increment = v_increment
+            first = 1
+            last = len(pressure_points)
+            step = 1
+        last_p_slice[pressure_axis] = slice(0, 1)
+        for p in range(first, last, step):
+            p_slice = [slice(None) for _ in range(data.ndim)]
+            p_slice[pressure_axis] = slice(p, p + 1)
+            data[p_slice] = np.where(
+                np.logical_or(
+                    data.mask[p_slice],
+                    np.logical_or(np.isnan(data[p_slice]), np.isinf(data[p_slice])),
+                ),
+                data[last_p_slice] + local_increment,
+                data[p_slice],
+            )
+            last_p_slice = p_slice
+
     @staticmethod
-    def fill_invalid(cube: Cube):
-        """Populate any invalid values with the maximum in that column on the assumption
-        that missing values fall below the model's surface."""
+    def _one_column_slice(cube: Cube) -> List[slice]:
+        """Create a slice object that will yield a single column from near the spatial centre
+        of the cube. Other leading dimensions will use the zeroth point."""
         (pressure_axis,) = cube.coord_dims("pressure")
-        cube_shape = np.array(cube.shape)
-        max_shape = cube_shape.copy()
-        max_shape[pressure_axis] = 1
-        max_values = np.nanmax(cube.data, axis=pressure_axis).reshape(max_shape)
-        cube.data = np.where(np.isnan(cube.data), max_values, cube.data)
+        x_point = cube.shape[-1] // 2
+        y_point = cube.shape[-2] // 2
+        used_point = min(x_point, y_point)
+        one_column = [slice(0, 1) for _ in range(cube.ndim)]
+        one_column[pressure_axis] = slice(None)
+        one_column[-2:] = [slice(used_point, used_point + 1)] * 2
```

Comment:
IMPROVER has a convention that this is always true, so I think this is reasonable.

Reply:
This has come up before on other PR's with the concern being that while this might be the case for us it isn't necessarilly true for collaborators or anyone else using the code. If someone used a cube with say ["latitude","longitude","pressure","realization"] as the coordinates the class might run but produce something very odd and it wouldn't be obvious why without looking very carefully through the code. 

Atleast a comment saying that this bit of the code should be indexing the x and y coordinate might make it easier for someone looking through at a later date.

Reply:
After a bit more thought, I realised that we no longer need to know where the x and y axes are, because the requirement for querying a single column has gone, now that I've passed the onus on specifying the variable behaviour to the calling programme. So this problem has gone away. I was convinced that I needed to do something about it though.